mockRestore does not restore the original implementation

### Describe the bug

According to the [docs](https://vitest.dev/api/mock.html#mockrestore) mockRestore should restore the original implementation. 

I would expect the following test to pass: 

```ts
import { expect, test, vi } from 'vitest';

const testFn = vi.fn(() => true);

test('.mockRestore() should restore initial implementation', () => {
  expect(testFn()).toBe(true);

  testFn.mockReturnValue(false);
  expect(testFn()).toBe(false);

  testFn.mockRestore();
  expect(testFn()).toBe(true);
});
```
I get the following error on stack blitz:

```
AssertionError: expected undefined to be true // Object.is equality
```


It seems that instead of restoring the implementation it's wiped completely. 
I'd expect that from [mockReset](https://vitest.dev/api/mock.html#mockreset) not mockRestore.

### Reproduction

[StackBlitz Reproduction](https://stackblitz.com/edit/vitest-dev-vitest-kgywe2?file=test%2Fbasic.test.ts,vite.config.ts&initialPath=__vitest__)

### System Info

```shell
System:
    OS: macOS 13.3.1
    CPU: (8) x64 Apple M1
    Memory: 25.78 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 3.3.0 - ~/.nvm/versions/node/v18.12.1/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
  Browsers:
    Chrome: 113.0.5672.92
    Safari: 16.4
  npmPackages:
    vitest: ^0.31.0 => 0.31.0
```


### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/spy/src/index.ts
+++ b/packages/spy/src/index.ts
@@ -324,7 +324,7 @@ export function fn<TArgs extends any[] = any[], R = any>(
 export function fn<TArgs extends any[] = any[], R = any>(
   implementation?: (...args: TArgs) => R,
 ): Mock<TArgs, R> {
-  const enhancedSpy = enhanceSpy(tinyspy.internalSpyOn({ spy: () => {} }, 'spy'))
+  const enhancedSpy = enhanceSpy(tinyspy.internalSpyOn({ spy: implementation || (() => {}) }, 'spy'))
   if (implementation)
     enhancedSpy.mockImplementation(implementation)
 

--- a/test/core/test/jest-mock.test.ts
+++ b/test/core/test/jest-mock.test.ts
@@ -322,7 +322,7 @@ describe('jest mock compat layer', () => {
     expect(Fn.mock.instances[1]).toBe(instance2)
   })
 
-  test('.mockRestore() should restore initial implementation', () => {
+  it('.mockRestore() should restore initial implementation', () => {
     const testFn = vi.fn(() => true)
     expect(testFn()).toBe(true)
 

--- a/test/core/test/jest-mock.test.ts
+++ b/test/core/test/jest-mock.test.ts
@@ -321,4 +321,15 @@ describe('jest mock compat layer', () => {
     const instance2 = new Fn()
     expect(Fn.mock.instances[1]).toBe(instance2)
   })
+
+  test('.mockRestore() should restore initial implementation', () => {
+    const testFn = vi.fn(() => true)
+    expect(testFn()).toBe(true)
+
+    testFn.mockReturnValue(false)
+    expect(testFn()).toBe(false)
+
+    testFn.mockRestore()
+    expect(testFn()).toBe(true)
+  })
 })

